broadinstitute/single_cell_portal_core
# Differential expression prototype UX (SCP-4272)

This prototypes a frontend and UX for new differential expression (DE) features.

### New feature overview
Previously, users had no _a priori_ way in SCP to find genes that are expressed at different levels in one group of cells compared to other groups of cells.  Users would have to search many genes, basically randomly, making such insight infeasible.  Now, leveraging SCP ingest infrastructure, Scanpy, and our UI, users can trivially find such differentially expressed genes with a few clicks.  

Like related genes ideogram helps explore by revealing genes with pathway- or evolutionary relationships to a searched gene, this new DE UX also suggests particular genes to explore -- but using data from strictly _within_ the study.  These orthogonal approaches often complement each other, enriching the ways we help users gain insight from single-cell studies.  Beyond introducing a new-to-SCP kind of data, the DE UX is unique in that it suggests genes to search whether already viewing a gene or not.

### Technical summary
Upon selecting a group for differential expression viewing, the frontend fetches pre-computed DE data from a Google bucket.  The data reports a handful of DE metrics for all genes in all cells of that group, compared to all groups _except_ the selected group -- which we often call "A vs. all" DE data.  The frontend parses the ~2 MB precomputed DE TSV file, and shows a table with log<sub>2</sub> fold-change and adjusted p-values for each gene, and in-app links to search the gene.  Clicking the gene searches it, showing detailed expression patterns with more granularity than the tabulated DE data.

### Limitations
This is a first draft.  It's hidden behind a feature flag.  Future work will refine the UX per recent demo feedback, and refine the frontend as documented in TODO comments.  These are tracked in another ticket.

### Demo

https://user-images.githubusercontent.com/1334561/166719715-04dcdf9f-61b5-4c2d-9de9-da8d88f542cc.mov

### Manual test instructions
A new test study gives a way to locally reproduce a real public dataset with easy-to-analyze DE patterns.

0. Load [human genome annotation in species metadata](https://broadinstitute.slack.com/archives/CBEHTH601/p1651603466458739?thread_ts=1651603449.817109&cid=CBEHTH601), if you haven't already
1.  Start delayed job runner: execute `rails jobs:work` in a terminal
2.  Load public test DE study: `echo "SyntheticStudyPopulator.populate('human_milk_de_49k_cell')" | bundle exec rails console -e development`.
3. In output of above, copy the `bucket_id` value
4. Copy the reference public test DE study's `_scp-internal` directory to your new study's bucket: `gsutil -m cp -r "gs://broad-singlecellportal-public/test/studies/SCP1671/_scp-internal" "gs://<the bucket_id value you copied from step 2>/"`
5.  Run migration: `rake db:migrate`
6.  Sign in to your local SCP, open menu at top-right of page, select "Feature flags"
7.  Find yourself, set `differential_expression_frontend` feature flag to `true`
8.  Go to Study Overview for your new public test DE study, "Human milk - differential expression"
9.  Go to "Settings", "View options", set default cluster to `All Cells UMAP` and default annotation to `cell_type__ontology_label`, click "Update settings"
10.  Go to Explore tab, refresh, click "Differential expression" button
11.  In resulting modal, select group "epithelial cell"
12. Confirm that a table of genes is shown in right panel
13. Click gene "CSN2" in DE panel
14. Confirm CSN2 gene is searched
15. If steps 11 or 13/14 fails, then it might be because ingest hasn't completed; try selecting group "macrophage" and searching gene "APLP2" instead.

This builds on SCP-4272, with the intent to get engineering feedback to inform the upcoming DE MVP.
